docs/proposals/service-catalog.md

```diff
@@ -0,0 +1,408 @@
+of resources as a unit.  Obviously, Compose does not create Kubernetes resources
+as Helm does, including Services which make the service offered by the
+composition discoverable.
+
+### OpenShift Templates
+
+Openshift templates offer a feature set similar to Helm.  Templates are
+resources created as yaml files and imported as a resource of type "Template"
+into Openshift.  Templates are processed with the "oc process" command, filling
+in fields from parameter list and can even dynamically generate parameters like
+passwords, for example. Once processed, the template becomes a single yaml file
+output that can be fed to `oc create -f` to create the resources in Openshift.
+
+The use case application is the same as that of Helm.  Once resources created by
+the template, they are not tracked any further as a member of the template.  It
+is up to the users to apply labels such that all resources created by a template
+can be selected in the cluster.
+
+### Cloud Foundry Service Broker
+
+The CF Service Broker is more full featured than the previous examples, in that
+it implements a catalog of services (CCDB) provided by any number of service
+brokers and claimable by cluster users.  The cloud controller can discover
+services from service brokers that implement it. the Service Broker API
+using the "catalog" call.  Other calls that happen in response to a user
+claiming the service from the catalog are "provision", "bind", "unbind", and
+"deprovision".  The separation of the provisioning stage from the binding stage
+allows for asynchronous dynamic provisioning of a service for a particular user.
+
+This satisfies all the use cases except for the syncing of changed credentials.
+There is no mechanism for the Service Broker to initiate a deprovision or
+resync.
+
+## Prototype
+
+We have implemented a prototype to demonstrate a possible workflow for Catalogs
+in a Kubernetes cluster.  The implementation is
+[here](https://github.com/sjenning/kubernetes/tree/catalog-apiserver).
+
+This prototype allows for the following type of workflow:
+
+1. A user (the “provider” or “publisher”) creates one or more resources that
+   he/she wants to share    with others. For this example, let’s imagine that I
+   want to share a secret I created with other    users.
+1. The publisher creates a catalog posting that references the secret.
+1. Another user (the “consumer”) searches for and locates the entry in the
+   catalog for the secret.
+1. The consumer instantiates the catalog entry (we’re currently calling this a
+   claim, but the name is open for discussion)
+1. The end result is a new secret in the consumer’s namespace.
+1. If the publisher changes the contents of the secret, all consumers’ secrets
+   are automatically updated with the latest data.
+
+We added a new executable that runs an apiserver and controllers related to
+catalogs. This process handles requests to a new API group,
```

Comment:
For others:
https://github.com/sjenning/kubernetes/blob/catalog-apiserver/pkg/apis/servicecatalog/v1alpha1/types.go